Add Ombi integration

larssont opened the pull request:
## Description:
Integration for Ombi sensors and services.

**Related issue (if applicable):** N/A

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#10400

## Example entry for `configuration.yaml` (if applicable):
```yaml
ombi:
  api_key: b5y5cxs3g8a6mskpmegdfdvvfc3o2wp4
  host: 192.168.1.120
  username: myusername42
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

Commit:
Add Ombi integration

Commit:
Black

Commit:
Remove trailing comma

MartinHjelmare commented on homeassistant/components/ombi/sensor.py:
```diff
@@ -0,0 +1,125 @@
+        vol.Optional(CONF_URLBASE, default=DEFAULT_URLBASE): cv.string,
+        vol.Optional(CONF_SSL, default=DEFAULT_SSL): cv.boolean,
+        vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL): cv.time_period,
+        vol.Optional(CONF_MONITORED_CONDITIONS, default=list(SENSOR_TYPES)): vol.All(
```
We don't allow monitored conditions in config anymore.

https://github.com/home-assistant/architecture/blob/master/adr/0003-monitor-condition-and-data-selectors.md

Please create all sensor types by default.

larssont replied:
Since the sensors make individual API calls, it would seem reasonable to allow the users to limit the number of requests by omitting certain sensors from monitored_conditions. Should I change it anyway? Ombi is a self-hosted service, so the number of API calls are usually not an issue.

MartinHjelmare replied:
If it's usually not an issue, I suggest we remove the option.

larssont replied:
Alright, fixed!

MartinHjelmare commented on homeassistant/components/ombi/sensor.py:
```diff
@@ -0,0 +1,125 @@
+
+    for condition in conditions:
+        sensor_label = condition
+        sensor_type = SENSOR_TYPES[condition].get("type")
```
Please use `dict[key]` if we know that the key is in the dict. Don't use `dict.get`.

MartinHjelmare commented on homeassistant/components/ombi/sensor.py:
```diff
@@ -0,0 +1,125 @@
+        """Update the sensor."""
+        try:
+            if self._label == "pending":
+                self._state = self._ombi.pending_requests
```
Is the ombi data updated when we access the `pending_requests` attribute?

larssont replied:
Yes, that's correct. The same goes for all other sensors, whenever an attribute is accessed a new request goes to the API.

MartinHjelmare commented on homeassistant/components/ombi/sensor.py:
```diff
@@ -0,0 +1,125 @@
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+        vol.Optional(CONF_URLBASE, default=DEFAULT_URLBASE): cv.string,
+        vol.Optional(CONF_SSL, default=DEFAULT_SSL): cv.boolean,
+        vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL): cv.time_period,
```
Please define the `SCAN_INTERVAL` constant at the module level instead. This will set the default scan interval of this platform.

larssont replied:
Should be fixed now.

MartinHjelmare reviewed (changes requested):
Code looks great!

Is there a chance that more platforms will be created for this integration, eg media_player, or custom services? In that case we should consider moving the config section to under the integration key and set up the sensor platform via discovery from the component module.

Commit:
Change dict.get to dict[key] for known keys

larssont commented:
@MartinHjelmare 
Alright, I looked over the API and did some tests and I think implementing some custom services would definitely be possible. Since Ombi's main purpose is to request movies and TV Shows and also approve requests, it would be feasible to implement a custom service that for instance searched and subsequently requested something on Ombi through the component. Although, I would have to update the pyombi package before implementing any of that here obviously.

Commit:
Remove monitored conditions from config

Commit:
Define SCAN_INTERVAL for default scan interval

MartinHjelmare commented:
I would recommend putting the config under the integration domain key instead and refactor to use `helpers.discovery.load_platform` to load the platform from the component module.

We need to define a `CONFIG_SCHEMA` in the component and remove the `PLATFORM_SCHEMA` in the platform.

The ombi api instance should be created in the component and stored in hass.data, to later be accessible from the platform.

tidusjar commented:
Nice!

Commit:
Adjust requests syntax and add music_requests

Commit:
Remove Ombi object initialization

Commit:
Move constants to const.py

Commit:
Fix imports

Commit:
Set pyombi requirement to version 0.1.3

Commit:
Add services.yaml

Commit:
Add config schema and setup for integration

Commit:
Set pyombi requirement to version 0.1.3

Commit:
Fix syntax for scan interval

Commit:
Fix datetime import

Commit:
Add all files from ombi component

Commit:
Move imports around

Commit:
Move imports around

larssont commented:
I think everything should be included now. I separated the config from the sensor setup and added a couple for services calls for different types of requests. I've added in music support as well just for the sake of it. The only thing I'm a bit indecisive about is if I should implement some type of response on how each service request was handled, i.e. whether they succeeded or not. Let me know if you have any ideas for this.

MartinHjelmare commented on homeassistant/components/ombi/__init__.py:
```diff
@@ -0,0 +1,96 @@
+
+def setup(hass, config):
+    """Set up the Ombi component platform."""
+    import pyombi
```
Please move 3rd party imports to the top of the module.

MartinHjelmare commented on homeassistant/components/ombi/__init__.py:
```diff
@@ -0,0 +1,96 @@
+    extra=vol.ALLOW_EXTRA,
+)
+
+SCAN_INTERVAL = timedelta(seconds=60)
```
We can remove this. It doesn't have a function in component modules. Define it in the platform instead.

MartinHjelmare commented on homeassistant/components/ombi/__init__.py:
```diff
@@ -0,0 +1,96 @@
+    """Set up the Ombi component platform."""
+    import pyombi
+
+    urlbase = f"{config[DOMAIN][CONF_URLBASE].strip('/') if config[DOMAIN][CONF_URLBASE] else ''}/"
```
This option has a default.

MartinHjelmare replied:
Please add a custom validator that strips the slash from the url. We want all config modifications to happen in the config schema. See `config_validation.py` for examples.

MartinHjelmare commented on homeassistant/components/ombi/__init__.py:
```diff
@@ -0,0 +1,96 @@
+        if music:
+            ombi.request_music(music[0]["foreignAlbumId"])
+
+    hass.services.register(DOMAIN, "submit_movie_request", submit_movie_request)
```
Please add service validation schemas for the services.

MartinHjelmare commented on homeassistant/components/ombi/sensor.py:
```diff
@@ -0,0 +1,72 @@
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Ombi sensor platform."""
+    sensors = []
```
Please add a guard clause here that checks if discovery_info is None and return if so. This will guard from setting up the platform via a platform config section.

MartinHjelmare commented:
We can raise an error in the service handler on failure. This will be shown to the user.

Commit:
Move pyombi import to top of module

Commit:
Move scan_interval to sensor module

Commit:
Add custom validator for urlbase

Commit:
Add guard clause for discovery_info

Commit:
Add service validation schemas and constants

Commit:
Bump pyombi version

Commit:
Adjust urlbase validation

Commit:
Add exception warnings for irretrievable media

Commit:
Bump pyombi

MartinHjelmare commented on homeassistant/components/ombi/__init__.py:
```diff
@@ -61,36 +92,58 @@ def setup(hass, config):
     def submit_movie_request(call):
         """Submit request for movie."""
-        name = call.data.get("name")
+        name = call.data.get(CONF_NAME)
```
Please use `dict[key]` for required service data keys and keys with default schema values.

MartinHjelmare commented on homeassistant/components/ombi/__init__.py:
```diff
@@ -0,0 +1,149 @@
+
+SUBMIT_TV_REQUEST_SERVICE_SCHEME = vol.Schema(
+    {
+        vol.Required(CONF_NAME): cv.string,
```
We only use `CONF_*` constants for config keys. For service data keys we use `ATTR_*` constants.

MartinHjelmare commented on homeassistant/components/ombi/__init__.py:
```diff
@@ -0,0 +1,149 @@
+    return value + "/"
+
+
+SUBMIT_MOVIE_REQUEST_SERVICE_SCHEME = vol.Schema({vol.Required(CONF_NAME): cv.string})
```
`SUBMIT_MOVIE_REQUEST_SERVICE_SCHEMA`

MartinHjelmare commented on homeassistant/components/ombi/__init__.py:
```diff
@@ -0,0 +1,149 @@
+        ombi.test_connection()
+    except pyombi.OmbiError as err:
+        _LOGGER.warning("Unable to setup Ombi: %s", err)
+        return
```
```py
return False
```

MartinHjelmare commented on homeassistant/components/ombi/sensor.py:
```diff
@@ -0,0 +1,78 @@
+        except OmbiError as err:
+            _LOGGER.warning("Unable to update Ombi sensor: %s", err)
+            self._state = None
+            return
```
Not needed `return`.

Commit:
Change from .get to dict[key]

Commit:
Change schema and conf variable names

Commit:
Set return to return false

Commit:
Remove unneeded return

MartinHjelmare reviewed (approved):
Looks great!

MartinHjelmare commented:
Please update the config example in the PR description.

larssont commented:
Great! Thanks for all the feedback, I'll make sure to update the docs.

MartinHjelmare commented:
Ping me when the yaml example above is updated and I'll merge here.

larssont commented:
@MartinHjelmare Fixed!
